src/utils/delegation.ts

```diff
@@ -0,0 +1,113 @@
+import { subgraphRequest } from '../utils';
+import delegationSubgraphs from '../delegationSubgraphs.json';
+
+export const SNAPSHOT_SUBGRAPH_URL = delegationSubgraphs;
+const PAGE_SIZE = 1000;
+
+type Delegation = {
+  delegator: string;
+  delegate: string;
+  space: string;
+  timestamp: number;
+};
+
+export default async function getDelegatesBySpace(
+  network: string,
+  space: string,
+  snapshot: string | number = 'latest',
+  options: any = {}
+) {
+  const subgraphUrl = options.subgraphUrl || SNAPSHOT_SUBGRAPH_URL[network];
+  if (!subgraphUrl) {
+    return Promise.reject(
+      `Delegation subgraph not available for network ${network}`
+    );
+  }
+
+  let result: Delegation[] = [];
+  const spaceIn = ['', space];
+  if (space.includes('.eth')) spaceIn.push(space.replace('.eth', ''));
+
+  while (true) {
+    const newResults = await fetchData({
+      url: subgraphUrl,
+      spaces: spaceIn,
+      pivot: result[result.length - 1]?.timestamp || 0,
+      snapshot
+    });
+
+    result = mergeWithoutDuplicates(result, newResults);
+
+    if (
+      newResults.length === PAGE_SIZE &&
+      newResults[0].timestamp === newResults[newResults.length - 1].timestamp
+    ) {
+      throw new Error('Unable to paginate delegation');
+    }
+
+    if (newResults.length < PAGE_SIZE) break;
+  }
+
+  return result;
+}
+
+export function mergeWithoutDuplicates(a: Delegation[], b: Delegation[]) {
```

Comment:
I just realized that subgraph returns `id` property, i think we can use this property to filter duplicates here than worrying about timestamps?

Reply:
the `id` field is somewhat expensive:

```
 {
        "id": "0x000000085d9a759bb5c3d459d638739c0f48deb0-stgdao.eth-0x7265a60acaeaf3a5e18e10bc1128e72f27b2e176",
        "delegator": "0x000000085d9a759bb5c3d459d638739c0f48deb0",
        "space": "stgdao.eth",
        "delegate": "0x7265a60acaeaf3a5e18e10bc1128e72f27b2e176",
        "timestamp": 1681033668
      },
```

We still need the `timestamp` property, and adding the `id` field will increase the amount of data, 35kb vs 43kb for 1000 delegations.

We decrease the amount of computation, at the cost of more data.

For cleaner duplication check, we could also use our own unique keys. So instead of using arrays, we could just use something like:

```
const results = new Map<string, Delegation>();

....

const newResults = await fetchData(....)

newResults.forEach((result: Delegation) => {
  const key = uniqueKey(result);
  if (!results.has(key)) results.set(key, result);
})

function uniqueKey(d: Delegation( {
  return `${d.delegator}-${d.delegate}-${d.space}`
}